bundles/org.openhab.binding.modbus.helioseasycontrols/src/main/java/org/openhab/binding/modbus/helioseasycontrols/internal/HeliosEasyControlsHandler.java

```diff
@@ -0,0 +1,577 @@
+
+        if (comms == null) {
+            @SuppressWarnings("null")
+            String label = Optional.ofNullable(getBridge()).map(b -> b.getLabel()).orElse("<null>");
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE,
+                    String.format("Bridge '%s' not completely initialized", label));
+            logger.debug("Bridge not initialized fully (no endpoint) -- aborting init for {}", this);
+            return;
+        }
+    }
+
+    @Override
+    public void initialize() {
+        this.config = getConfigAs(HeliosEasyControlsConfiguration.class);
+        this.readVariableDefinition();
+        this.connectEndpoint();
+        if (this.comms != null) {
+            HeliosEasyControlsHandler.transactionLocks.putIfAbsent(this.comms.getEndpoint(), new Semaphore(1, true));
+            updateStatus(ThingStatus.UNKNOWN);
+
+            // background initialization
+            scheduler.execute(() -> {
+                try {
+                    readValue(HeliosEasyControlsBindingConstants.DATE_FORMAT);
+                    // status will be updated to ONLINE by the read callback function
+                } catch (HeliosException e) {
+                    logger.error(e.getMessage());
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                }
+            });
+
+            // poll for status updates regularly
+            Runnable runnable = new Runnable() {
+
+                @Override
+                public void run() {
+                    for (Map.Entry<String, HeliosVariable> entry : vMap.entrySet()) {
+                        if (isLinked(entry.getKey())
+                                || (entry.getKey().equals(HeliosEasyControlsBindingConstants.DATE_FORMAT))) {
+                            try {
+                                readValue(entry.getKey());
+                            } catch (HeliosException e) {
+                                logger.error(e.getMessage());
+                                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                        e.getMessage());
+                            }
+                        }
+                    }
+                }
+            };
+            this.pollingJob = scheduler.scheduleAtFixedRate(runnable, config.getRefreshInterval(),
+                    config.getRefreshInterval(), TimeUnit.MILLISECONDS);
+
+        } else {
+            logger.error("Modbus endpoint is null.");
+        }
+    }
+
+    @Override
+    public void dispose() {
+        if (this.pollingJob != null) {
+            this.pollingJob.cancel(true);
+        }
+        this.comms = null;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        String channelId = channelUID.getId();
+        if (command instanceof RefreshType) {

```

Comment:
It's great that user can refresh the data on demand. Unfortunately, I think openHAB commands all channels REFRESH on startup, generating a lot of traffic since there are so many channels.

One way to handle this is to throttle refreshes. See FMI weather binding as one example https://github.com/openhab/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/AbstractWeatherHandler.java

Reply:
readValue uses the locking mechanism, so the traffic communicating with the device is handled in a controlled way. Or do you mean traffic in general?

Reply:
I think it would mean all these commands would pile up in queue of sort (=waiting for the lock), executed in very rapid fashion.

In my experience with other bindings it makes the openHAB itself slow (raspberry pi) since so many state updates are fired... Naturally after this initialization things are running smoothly again. So it's not a dealbreaker, just inconvenience 

You are correct that locking ensures that at most one transaction is running at the same time. But still the rate of the requests is quite high. The rate is limited by the parameters set in the tcp thing.

Writing this I realized that data is read by one variable at a time, not as one chunk. So actually my advice of throttling (I.e. Refreshing only if last refresh for the thing is old enough) does not make sense -- only way to refresh all the N channels is to fire N requests?

Here are instructions how to test it if you like: 
- create things 
- link all channels 
- start openHAB, observe if there are requests flying to the device (e.g. via logs). Check long it takes to update all channels 


I believe channels might get REFRESH when the things or items are recreated. This happens when the thigns/items files are modified.

Reply:
Exactly, the process is to write the variable number to a register and retrieve data from the register in return. Variable by variable.

At the moment the implementation will fetch all variables each refresh interval; one improvement could be to only fetch variables that are actually tied to an item. But so far when testing I didn't experience any particular delays (at least on my PC as well as on the Linux server running my PROD instance). Sure, some variable will take a little while to show a value in the sitemap, but I personally didn't mind so far...

Reply:
Actually, ignore my last statement 😂
I'm already just fetching the variables linked to an item...

Thinking about it, I'm wondering if simply introducing an additional delay before unlocking after a read transaction would have the same result as the throttling mechanism...

Btw. your choice of variable name for the throttling is quite amusing for an Austrian 😉 (cf. https://www.dict.cc/german-english/Fut.html)

Reply:
Thanks, then I'll leave the implementation as it is. If some users do experience issues I can still revisit this. But I wouldn't expect it to be honest.

Glad I could help with my little 'lesson in Austrian' 😁